### PR TITLE
Lower poly mul to standard

### PR DESCRIPTION
**Pull request by j2kun**

### CODE REVIEW

**Comment by j2kun** [comment]
Note to self:

```
    auto tensor = b.create<arith::ConstantOp>(DenseElementsAttr::get(
        tensorType, b.getIntegerAttr(tensorType.getElementType(), 0L)));
```

**Comment by j2kun** [comment]
Currently based off the branch in https://github.com/google/heir/pull/185

**Comment by j2kun** [comment]
Currently ready for a first-pass review, and rebased off of both

- https://github.com/google/heir/pull/192
- https://github.com/google/heir/pull/193

File: lib/Conversion/PolyToStandard/PolyToStandard.cpp [hunk 1]
```diff
@@ -721,7 +726,8 @@ void PolyToStandard::runOnOperation() {
 
   // target.addIllegalDialect<PolyDialect>();
   target.addIllegalOp<FromTensorOp, ToTensorOp, AddOp, SubOp, LeadingTermOp,
-                      MonomialOp, MonomialMulOp, ConstantOp, MulOp>();
+                      MonomialOp, MulScalarOp, MonomialMulOp, ConstantOp,
```
Comment (j2kun) [inline lib/Conversion/PolyToStandard/PolyToStandard.cpp:729]: This was the big problem: it wasn't in the illegal ops list, so it wasn't being lowered >_<

**Comment by j2kun** [comment]
This PR is a monster. I can split it into more manageable pieces after the conference this week.

**Comment by j2kun** [comment]
Pulled out these PRs:

- https://github.com/google/heir/pull/195
- https://github.com/google/heir/pull/196
- https://github.com/google/heir/pull/197

**Comment by j2kun** [comment]
@asraa  Integrated the fixes from our bug hunting session earlier today:

- Fixed bad cmpi predicate type in the lowering for leading_term (ne -> eq)
- Adding the missing addi op in the linalg generic for the naive polymul
- Updated tests

As mentioned, let's merge this in and then iterate on two things:

- Add a mod op to the linalg generic so that the accumulation can stay within the cmod at every step, which should simplify some of the computation and correctness issues in the remainder function
- Add robust mlir-cpu-runner testing of the lowering, started in https://github.com/google/heir/pull/203

File: lib/Conversion/PolyToStandard/PolyToStandard.cpp [hunk 2]
```diff
@@ -3,15 +3,19 @@
...
 #include "lib/Conversion/Utils.h"
+#include "llvm/include/llvm/ADT/TypeSwitch.h"           // from @llvm-project
+#include "llvm/include/llvm/Support/FormatVariadic.h"   // from @llvm-project
 #include "mlir/include/mlir/Dialect/Arith/IR/Arith.h"   // from @llvm-project
 #include "mlir/include/mlir/Dialect/Func/IR/FuncOps.h"  // from @llvm-project
 #include "mlir/include/mlir/Dialect/Func/Transforms/FuncConversions.h"  // from @llvm-project
-#include "mlir/include/mlir/Dialect/Linalg/IR/Linalg.h"  // from @llvm-project
-#include "mlir/include/mlir/Dialect/SCF/IR/SCF.h"        // from @llvm-project
-#include "mlir/include/mlir/Dialect/Tensor/IR/Tensor.h"  // from @llvm-project
+#include "mlir/include/mlir/Dialect/LLVMIR/LLVMDialect.h"  // from @llvm-rpoject
```
Comment (j2kun) [inline lib/Conversion/PolyToStandard/PolyToStandard.cpp:11]: note to self: fix

File: lib/Conversion/PolyToStandard/PolyToStandard.cpp [hunk 3]
```diff
@@ -367,47 +374,380 @@ struct ConvertAdd : public OpConversionPattern<AddOp> {
...
+  DenseMap<std::pair<Type, RingAttr>, func::FuncOp> modImpls;
+};
 
-    // TODO(https://github.com/google/heir/issues/143): Handle tensor of polys.
-    if (failed(applyPartialConversion(module, target, std::move(patterns)))) {
-      signalPassFailure();
-    }
+void PolyToStandard::generateOpImplementations() {
+  ModuleOp module = getOperation();
+  module.walk([&](Operation *op) {
```
Comment (asraa) [inline lib/Conversion/PolyToStandard/PolyToStandard.cpp:508]: I think here you can iterate directly on the MulOps.
```
getOperation().walk([&](MulOp op) {
```

File: lib/Conversion/PolyToStandard/PolyToStandard.cpp [hunk 4]
```diff
@@ -367,47 +374,380 @@ struct ConvertAdd : public OpConversionPattern<AddOp> {
...
+            b.getIndexType(), inputType.getElementType(), remainder);
+        auto cmpOp = b.create<arith::CmpIOp>(
+            arith::CmpIPredicate::sge, remainderLt.getDegree(), divisorDeg);
+        b.create<scf::ConditionOp>(cmpOp, remainder);
+      },
+      /*afterBuilder=*/
+      [&](OpBuilder &nestedBuilder, Location nestedLoc, ValueRange args) {
+        ImplicitLocOpBuilder b(nestedLoc, nestedBuilder);
+        Value remainder = args[0];
+        // TODO: move this out of the loop when it has ConstantLike trait
```
Comment (asraa) [inline lib/Conversion/PolyToStandard/PolyToStandard.cpp:656]: nit: link https://github.com/google/heir/issues/97

**Review by asraa — APPROVED** [review]
WOW! Thank you for the organization of the polymod function into a separate block, this made it 10x easier to understand.

**Pull request merged**